consent/strategy_default.go

```diff
@@ -1178,3 +1232,96 @@ func (s *DefaultStrategy) loginSessionFromCookie(r *http.Request) *flow.LoginSes
+	verifier := strings.Replace(uuid.New(), "-", "", -1)
+	challenge := strings.Replace(uuid.New(), "-", "", -1)
+	csrf := strings.Replace(uuid.New(), "-", "", -1)
+
+	// Generate the request URL
+	iu := s.getDeviceVerificationPath(ctx)
+	iu.RawQuery = r.URL.RawQuery
+
+	deviceRequest := &flow.DeviceUserAuthRequest{
+		ID:          challenge,
+		Verifier:    verifier,
+		CSRF:        csrf,
+		RequestURL:  iu.String(),
+		RequestedAt: time.Now().Truncate(time.Second).UTC(),
+	}
+	f, err := s.r.ConsentManager().CreateDeviceUserAuthRequest(
+		r.Context(),
+		deviceRequest,
+	)
+	if err != nil {
+		return errorsx.WithStack(err)
+	}
+
+	encodedFlow, err := f.ToDeviceChallenge(ctx, s.r)
+	if err != nil {
+		return err
+	}
+	store, err := s.r.CookieStore(ctx)
+	if err != nil {
+		return err
+	}
+
+	CookieNameDeviceCSRF := s.r.Config().CookieNameDeviceCSRF(ctx)
+	if err := createCsrfSession(w, r, s.r.Config(), store, CookieNameDeviceCSRF, csrf, s.c.ConsentRequestMaxAge(ctx)); err != nil {
+		return errorsx.WithStack(err)
+	}
+
+	query := url.Values{"device_challenge": {encodedFlow}}
+	if r.URL.Query().Has("user_code") {
+		query.Add("user_code", r.URL.Query().Get("user_code"))
+	}
+
+	http.Redirect(w, r, urlx.SetQuery(s.c.DeviceVerificationURL(ctx), query).String(), http.StatusFound)
+
+	// generate the verifier
+	return errorsx.WithStack(ErrAbortOAuth2Request)
+}
+
+func (s *DefaultStrategy) verifyDevice(ctx context.Context, _ http.ResponseWriter, r *http.Request, verifier string) (_ *flow.DeviceFlow, err error) {
+	ctx, span := trace.SpanFromContext(ctx).TracerProvider().Tracer("").Start(ctx, "DefaultStrategy.verifyAuthentication")
+	defer otelx.End(span, &err)
+
+	// We decode the flow from the cookie again because VerifyAndInvalidateDeviceRequest does not return the flow
+	f, err := flowctx.Decode[flow.DeviceFlow](ctx, s.r.FlowCipher(), verifier, flowctx.AsDeviceVerifier)
+	if err != nil {
+		return nil, errorsx.WithStack(fosite.ErrAccessDenied.WithHint("The device verifier is invalid."))
+	}
+
+	session, err := s.r.ConsentManager().VerifyAndInvalidateDeviceUserAuthRequest(ctx, verifier)
+	if errors.Is(err, sqlcon.ErrNoRows) {
+		return nil, errorsx.WithStack(fosite.ErrAccessDenied.WithHint("The device verifier has already been used, has not been granted, or is invalid."))
+	} else if err != nil {
+		return nil, err
+	}
+
+	if session.HasError() {
+		session.Error.SetDefaults(flow.DeviceRequestDeniedErrorName)
+		return nil, errorsx.WithStack(session.Error.ToRFCError())
+	}
+
+	store, err := s.r.CookieStore(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	CookieNameDeviceCSRF := s.r.Config().CookieNameDeviceCSRF(ctx)
+	if err := validateCsrfSession(r, s.r.Config(), store, CookieNameDeviceCSRF, session.Request.CSRF); err != nil {
+		return nil, err
+	}
+
+	return f, nil
+}
+
+func (s *DefaultStrategy) getDeviceVerificationPath(ctx context.Context) *url.URL {
+	return urlx.AppendPaths(s.c.PublicURL(ctx), "/oauth2/device/verify")
```

Comment:
dunno what framework it's been used, but if the handler is "named" u might be able to fetch it via a helper function, anyway like this looks good enough

Reply:
actually https://github.com/canonical/hydra/pull/3/files#diff-dc5158991c1ce571f0a59c32b86d6fc64c55fbe11d67cdee44b1b0ae436569dfR67

Reply:
I tried using that, the problem is that it creates a circular import (probably why they hard code the paths in https://github.com/canonical/hydra/blob/canonical/driver/config/provider.go#L468)